qualtran/bloqs/comparison_gates.ipynb

```diff
@@ -0,0 +1,957 @@
+    "\n",
+    "<p style=\"text-align: center;\"><a href=\"mailto:noureldinyosri@gmail.com\">Noureldin Yosri</a></p>\n",
+    "<p style=\"text-align: center;\">May 2023</p>"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "FS_qGUVPouBf"
+   },
+   "source": [
+    "## Abstract\n",
+    "Quantum comparison gates can be split into two categories: quantum-classical when comparing a quantum number to a classical number and quantum-quantum when comparing two quantum numbers.\n",
+    "\n",
+    "Comparison gates are an important building block for quantum algorithms, yet they are always relegated to the appendicies usually with figures showing how they work only on a small number of qubits. This leads to cases where researchers potentially waste time reinvents on already existing methods.\n",
+    "\n",
+    "The intention of this notebook is to serve as documentary of the current state of the art for comparison gates while expalining how they work both algorithmically and programmatically.\n",
+    "\n",
+    "**Note:** Most of the ideas explained here are built into [cirq-ft](https://github.com/quantumlib/Cirq/tree/master/cirq-ft). [Cirq](https://github.com/quantumlib/Cirq/)'s fault tolerant sublibrary."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "_-GG5ry9o6RI"
+   },
+   "source": [
+    "## Introduction\n",
+    "The current optimal implementation of a reversible oracle for comparing two quantum numbers of $n$ qubits each is $8n + \\mathcal{O}(1)$ and is given in the supplementary materials of [Berry et al., 2018](https://doi.org/10.1038/s41534-018-0071-5). Their implementation uses divide and conquer technique to create a binary tree of depth $\\log_2{(n)}$ whose leafs are the qubits of the numbers with intermediate values stored in $\\mathcal{O}(n)$ ancillas. Each non leaf node uses two CSWAP operations and the value of the comparison is computed from the root node using one Toffoli. \n",
+    "\n",
+    "When this decomposition is applied to the case where one input is a list of qubits and the the other is a classical number (i.e. a string of bits) this decomposition gives a $6n + \\mathcal{O}(1)$ T complexity since at the level right above the leafs half of the CSWAP operations degenerate to either identity or CNOTs, thus the number of CSWAPs becomes $\\approx \\frac{n}{2} + 2 \\times \\frac{n}{2} = \\frac{3n}{2}$, leadining to $4 \\times \\frac{3n}{2} = 6n$ T operations.\n",
+    "\n",
+    "However this is not the optimal way -in terms of T count- to compare a quantum number to a classical number which is $4n + \\mathcal{O}(1)$ T operations. The way to do can be found in appendix H of [Berry et al., 2019](\n",
+    "https://doi.org/10.48550/arXiv.1902.02134) and is based on reducing the problem to subtraction. The reduction in T count comes at the cost of linear depth as opposed to the logarithmic depth from the $6n$ decomposition.\n",
```

Comment:
Might be worth adding a short sentence how in the surface code, we're limited by T count and not depth

Reply:
added